BillsBuddy: Final app

fobadara opened the pull request:
In this branch I:
- created a budget app with rails
- added devise and cancancan for authentication and authorisation
- created model, request and integrated specs

fobadara committed:
Create model specs for user, group and bill

fobadara committed:
Merge pull request #3 from fobadara/2-model_testing

Bills Buddy: create model specs for user, group and bill

fobadara committed:
Create request specs for user, group and bills controller

fobadara committed:
Fix linter errors

fobadara committed:
Merge pull request #4 from fobadara/3-request_testing

Bills Buddy: request specs

fobadara committed:
Create integrated specs for bill, group, and login pages

fobadara committed:
Install capybra and update spec files

fobadara committed:
Fix linter errors

fobadara committed:
Merge pull request #5 from fobadara/4-integration_tests

Bills Buddy:  integration tests

fobadara committed:
Add authentication with devise

fobadara committed:
Merge pull request #6 from fobadara/5-authentication_with_devise

Add authentication with devise

fobadara committed:
Create models and controllers

fobadara committed:
Adjust code to specs

fobadara committed:
Update Readme file

fobadara committed:
Update Readme.md

fobadara committed:
Fix styling's linter errors

Davidosky007 commented:
Hi @fobadara, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

## You have submitted a project, that has Node_modules files on your repo making it difficult to perform a proper review also i get an error when i run the app locally. dont forget to provide a link to a video presentation 👍 
![image](https://user-images.githubusercontent.com/52497246/173202175-dccd3eb6-ab92-44c7-8c9a-99ebd03c9dcf.png)


Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

fobadara committed:
Reinstall gems

fobadara committed:
update package.json

fobadara committed:
update package.json

fobadara committed:
Update routes

fobadara committed:
Update procfile

fobadara committed:
Update procfile

fobadara committed:
Update procfile

fobadara committed:
Update procfile

fobadara committed:
Recreate Procfile

fobadara committed:
Update puma.rb

fobadara committed:
Recreate Procfile

fobadara committed:
Update gem file

fobadara committed:
Update procfile

fobadara committed:
Upadate package-lock file

fobadara committed:
Upadate package-lock file

fobadara committed:
Create Procfile

fobadara committed:
Update environment.rb

fobadara committed:
Include application.css

fobadara committed:
Include application.css

fobadara committed:
Include application.js to manifest

fobadara committed:
Update README.md

fobadara committed:
Fix linter errors

fobadara committed:
Fix linter errors

fobadara committed:
Fix linter errors

fobadara committed:
Update linters

fobadara committed:
Fix linter errors

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Update package-lock file

fobadara committed:
Fix system type error

fobadara commented:
Hello @Davidosky007,
Thanks for the review. I fixed all errors and added a live demo and the Heroku links to the top of my README.md file. It now works locally and online. You can run the server with "./bin/dev"

fobadara committed:
Update README.md

fobadara committed:
Update README.md

Davidosky007 commented on README.md:
```diff
@@ -1,24 +1,246 @@
+    <a href="https://github.com/fobadara/billsbuddy#readme"><strong>Explore the docs »</strong></a>
+    <br /> 
+    <br />
+    <a href="https://drive.google.com/file/d/1cb_558Y_uxh1vsqGh0YQumVN5pZKZgLZ/view?usp=sharing">View Demo</a>
```
[OPTIONAL] - KIndly change the title of this link to show it is for the video presentation

Davidosky007 commented on app/assets/stylesheets/custom.css:
```diff
@@ -0,0 +1,114 @@
+  gap: 0.6em;
+  padding: 1em;
+  height: 200px;
+  overflow-y: auto;
+  overflow-x: hidden;
+}
+
+.item {
+  padding: 1em;
+  border: 1px solid #eee;
+  border-radius: 20%;
+  background: rgb(241, 84, 17);
+  box-shadow: 0 0 6px #eee;
```
* Kindly stick to design provided as it is a requirement 
* Note these includes Colour, typogragphy, font, weight, layouts, spacing etc [Please refer to the rpoject requirements](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=You%20should%20follow,space%20between%20elements.)

fobadara replied:
The style for small screens now fits with the design provided

Davidosky007 commented on app/controllers/groups_controller.rb:
```diff
@@ -0,0 +1,85 @@
+    end
+  end
+
+  # # GET /groups/1/edit
+  # def edit
+  # end
+
+  # # POST /groups or /groups.json
+  # def create
+  #   @group = Group.new(group_params)
+
+  #   respond_to do |format|
+  #     if @group.save
+  #       format.html { redirect_to group_url(@group), notice: "Group was successfully created." }
+  #       format.json { render :show, status: :created, location: @group }
+  #     else
+  #       format.html { render :new, status: :unprocessable_entity }
+  #       format.json { render json: @group.errors, status: :unprocessable_entity }
+  #     end
+  #   end
+  # end
+
+  # # PATCH/PUT /groups/1 or /groups/1.json
+  # def update
```
* Kindly remove all commented code as it makes your codebase look bulky and unprofessional

Davidosky007 commented on app/models/user.rb:
```diff
@@ -0,0 +1,21 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
```
* As mentioned above please remove all commented code

Davidosky007 commented on db/migrate/20220530155039_devise_create_users.rb:
```diff
@@ -0,0 +1,44 @@
+      t.datetime :remember_created_at
+
+      ## Trackable
+      # t.integer  :sign_in_count, default: 0, null: false
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
```
* As mentioned try to remove all commented code

Davidosky007 commented on spec/features/group_show_spec.rb:
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe 'Group show', type: :feature do
+  before(:each) do
+    visit user_session_path
+
+    valid_image = Rack::Test::UploadedFile.new("#{Rails.root}/app/assets/images/test-logo.png")
+
+    @user = User.create(name: 'Tested User', email: 'test@mail.com', password: 'foobar123##')
+    @group1 = Group.create(name: 'Test Group', image: valid_image, user_id: @user.id)
+
+    @bill1 = Bill.create(name: 'Test Bill1', amount: 100, category: @group1, due_date: Date.today, user_id: @user.id)
+    @bill2 = Bill.create(name: 'Test Bill2', amount: 200, category: @group1, due_date: Date.today, user_id: @user.id)
+
+    fill_in 'Email', with: @user.email
+    fill_in 'Password', with: @user.password
+    click_button 'Log in'
+
+    visit group_path(@group1)
+  end
```
* Kindly fix your tests as 2 are failing, below ⬇️ is a screenshot 
![image](https://user-images.githubusercontent.com/52497246/173336974-cf8f34db-93e5-4780-ba02-d1ab5086a456.png)

Davidosky007 reviewed (changes requested):
Hi @fobadara 👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
**  Good job with all the changes you have made 👏 
** Nice video presentation 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

fobadara committed:
Update app setp instruction in README.md

fobadara committed:
Restyle app

fobadara committed:
Implement required changes

fobadara commented:
Hello @Davidosky007,
I am done with the required changes.

Davidosky007 reviewed (approved):
Hi @fobadara 👋 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


![job-welldone-gif](https://user-images.githubusercontent.com/52497246/173660801-0eacb360-34b6-431c-8842-c258e05976cb.gif)

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
